util/topgen.py

```diff
@@ -663,32 +663,62 @@ def generate_top_ral(top: Dict[str, object],
     # Collect up the memories to add
     mems = []
     for item in list(top.get("memory", [])):
-        byte_write = ('byte_write' in item and
-                      item["byte_write"].lower() == "true")
-        data_intg_passthru = ('data_intg_passthru' in item and
-                              item["data_intg_passthru"].lower() == "true")
-        size_in_bytes = int(item['size'], 0)
-        num_regs = size_in_bytes // addrsep
-        swaccess = access.SWAccess('top-level memory',
-                                   item.get('swaccess', 'rw'))
-
-        mems.append(window.Window(name=item['name'],
-                                  desc='(generated from top-level)',
-                                  unusual=False,
-                                  byte_write=byte_write,
-                                  data_intg_passthru=data_intg_passthru,
-                                  validbits=regwidth,
-                                  items=num_regs,
-                                  size_in_bytes=size_in_bytes,
-                                  offset=int(item["base_addr"], 0),
-                                  swaccess=swaccess))
+        mems.append(create_mem(item, item['name'], addrsep, regwidth))
+
+    # Top-level may override the mem setting. Store the new type to name_to_block
+    # If no other instance uses the orignal type, delete it
+    list_orginal_types = []
+    for module in top['module']:
+        if 'memory' in module.keys() and len(module['memory']) > 0:
+            block = deepcopy(name_to_block[module['type']])
+            name_to_block[module['name']] = block
+            inst_to_block[module['name']] = module['name']
+
+            if module['type'] not in list_orginal_types:
+                list_orginal_types.append(module['type'])
+
+            for mem_name, item in module['memory'].items():
+                assert block.reg_blocks[mem_name]
+                assert len(block.reg_blocks[mem_name].windows) <= 1
+
+                win = create_mem(item, mem_name, addrsep, regwidth)
+                if len(block.reg_blocks[mem_name].windows) > 0:
+                    block.reg_blocks[mem_name].windows[0] = win
+                else:
+                    block.reg_blocks[mem_name].windows.append(win)
```

Comment:
When does this case happen? To me, it looks like it would mean a mismatch between the block (which thinks it has no memory regions) and the chip-level configuration. Can we just make it an error?

Reply:
for sram_ctrl, we actual don't have memory in the block and we define the memory in chip-level for sram_ctrl_main and sram_ctrl_ret. for detail, see #7218

Reply:
indeed, we currently have both cases: 
- blocks that define a window since the memory size is fixed at the block level (e.g. `rom_ctrl`)
- blocks that do not define a window at the block level, since the window size is not fixed and needs to be set at the top level (e.g. `sram_ctrl`).

btwy, should we add a check here that ensures the top-level configuration actually matches the window size in case a window is defined at the IP level?

Reply:
Oh, I see. Now I understand.

Good idea, @msfschaffner, about checking consistency if both are defined. @weicaiyang: could you add that please?